superduperdb/components/datatype.py

```diff
@@ -204,148 +233,212 @@ def get_encodable_cls(cls, name, default=None):
         Get the subclass of the _BaseEncodable with the given name.
         All the registered subclasses must be subclasses of the _BaseEncodable.
         """
-        for sub_cls in cls.__subclasses__():
-            if sub_cls.__name__ == name:
+        for sub_cls in _find_descendants(cls):
+            if sub_cls.__name__.lower() == name.lower().replace('_', '').replace(
+                '-', ''
+            ):
                 return sub_cls
         if default is None:
             raise ValueError(f'No subclass with name "{name}" found.')
         elif not issubclass(default, cls):
             raise ValueError(
                 "The default class must be a subclass of the _BaseEncodable."
             )
-
         return default
 
+    @classmethod
+    @abstractmethod
+    def _get_object(cls, db, r):
+        pass
+
+    @classmethod
+    @abstractmethod
+    def decode(cls, r, db=None) -> '_BaseEncodable':
+        pass
+
 
 @dc.dataclass
 class Encodable(_BaseEncodable):
-    def encode(
-        self,
-        bytes_encoding: t.Optional[BytesEncoding] = None,
-        leaf_types_to_keep: t.Sequence = (),
-    ) -> t.Union[t.Optional[str], t.Dict[str, t.Any]]:
-        """
-        :param bytes_encoding:
-        """
-        from superduperdb.backends.base.artifact import ArtifactSavingError
-
-        def _encode(x):
-            try:
-                bytes_ = self.datatype.encoder(x)
-            except Exception as e:
-                raise ArtifactSavingError from e
-            sha1 = str(hashlib.sha1(bytes_).hexdigest())
-            if (
-                CFG.bytes_encoding == BytesEncoding.BASE64
-                or bytes_encoding == BytesEncoding.BASE64
-            ):
-                # TODO: artifact stores is not compatible with non-bytes types
-                bytes_ = to_base64(bytes_)
+    x: t.Optional[t.Any] = None
+    leaf_type: t.ClassVar[str] = 'encodable'
 
-            return bytes_, sha1
+    def __post_init__(self):
+        if self.uri is not None and self.file_id is None:
+            self.file_id = _construct_file_id_from_uri(self.uri)
 
-        if self.datatype.encoder is None:
-            return self.x
+    def _encode(self):
```

Comment:
A bug I found today about this.

Now  there are to model

model1.predict_kwargs = {"threshold": 0.5}
model2.predict_kwargs = {"threshold": 0.5}


If I set predict_kwargs as the artifact, these two artifacts will have the same `sha1`, and then conflicts occur when artifacts are deleted, updated, etc.

For example, If we deleted model1, model2 will lost the aritfact.

Reply:
I see